source-dropbox/oauth2.go

```
@@ -0,0 +1,141 @@
+
+	scopes := GetScopes()
+
+	spec := pf.OAuth2{
+		Provider:                  "dropbox",
+		AuthUrlTemplate:           fmt.Sprintf(AUTH_URL_TEMPLATE_FORMAT_STR, strings.Join(scopes, " ")),
+		AccessTokenUrlTemplate:    "https://api.dropboxapi.com/oauth2/token",
+		AccessTokenHeadersJsonMap: map[string]json.RawMessage{"Content-Type": json.RawMessage(`"application/x-www-form-urlencoded"`)},
+		AccessTokenBody:           "client_id={{{ client_id }}}&client_secret={{{ client_secret }}}&code={{{ code }}}&grant_type=authorization_code&redirect_uri={{{ redirect_uri }}}",
+		AccessTokenResponseJsonMap: map[string]json.RawMessage{
+			"access_token":  json.RawMessage(`"/access_token"`),
+			"refresh_token": json.RawMessage(`"/refresh_token"`),
+			"token_type":    json.RawMessage(`"/token_type"`),
+			"expires_in":    json.RawMessage(`"/expires_in"`),
+			"scope":         json.RawMessage(`"/scope"`),
+			"uid":           json.RawMessage(`"/uid"`),
+			"account_id":    json.RawMessage(`"/account_id"`),
+		},
+	}
+
+	return &spec
+}
+
+type Credentials struct {
+	AccessToken  string `json:"access_token"`
+	RefreshToken string `json:"refresh_token"`
+	TokenType    string `json:"token_type"`
+	ExpireTime   int    `json:"expires_in"`
+	Scope        string `json:"scope"`
+	UID          string `json:"uid"`
+	AccountID    string `json:"account_id"`
+	ClientID     string `json:"client_id"`
+	ClientSecret string `json:"client_secret"`
+}
+
+func (c *Credentials) Validate() error {
+	requiredProperties := [][]string{
+		{"AccessToken", c.AccessToken},
+	}
+	for _, req := range requiredProperties {
+		if req[1] == "" {
+			return fmt.Errorf("missing '%s'", req[0])
+		}
+	}
+	return nil
+}
+
+func (c *Credentials) GetOauthClient(ctx context.Context) (*http.Client, error) {
```

This function doesn't need to return an error